airbyte-integrations/connectors/source-chargebee/source_chargebee/streams.py

```diff
@@ -0,0 +1,172 @@
+from chargebee.model import Model
+from chargebee.models import Addon, Customer, Invoice, Order, Plan, Subscription
+
+# Backoff params below
+# according to Chargebee's guidance on rate limit
+# https://apidocs.chargebee.com/docs/api?prod_cat_ver=2#api_rate_limits
+MAX_TRIES = 10  # arbitrary max_tries
+MAX_TIME = 90  # because Chargebee API enforce a per-minute limit
+
+
+class ChargebeeStream(Stream):
+    primary_key = "id"
+    cursor_field = "updated_at"
+    page_size = 100
+    api: Model = None
+
+    def __init__(self, start_date: str):
+        # Convert `start_date` to timestamp(UTC).
+        self._start_date = int(datetime.strptime(start_date, "%Y-%m-%dT%H:%M:%SZ").replace(tzinfo=timezone.utc).timestamp())
+
+    @property
+    def state_checkpoint_interval(self) -> int:
+        return self.page_size
+
+    def next_page_token(self, list_result: ListResult) -> Optional[Mapping[str, Any]]:
+        # Reference for Chargebee's pagination strategy below:
+        # https://apidocs.chargebee.com/docs/api/#pagination_and_filtering
+        next_offset = list_result.next_offset
+        if next_offset:
+            return {"offset": next_offset}
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        params = {
+            "limit": self.page_size,
+            "include_deleted": "true",
+            "sort_by[asc]": self.cursor_field,
+        }
+
+        if next_page_token:
+            params.update(next_page_token)
+
+        start_point = self._start_date
+        if stream_state:
+            start_point = max(start_point, stream_state[self.cursor_field])
+
+        params[f"{self.cursor_field}[after]"] = start_point
+
+        return params
+
+    def parse_response(self, list_result: ListResult, **kwargs) -> Iterable[Mapping]:
+        for message in list_result:
+            yield message._response[self.name]
+
+    @backoff.on_exception(
+        backoff.expo,  # Exponential back-off
+        OperationFailedError,  # Only on Chargebee's OperationFailedError
+        max_tries=MAX_TRIES,
+        max_time=MAX_TIME,
+    )
+    def _send_request(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> ListResult:
+        """
+        Just a wrapper to allow @backoff decorator
+        Reference: https://apidocs.chargebee.com/docs/api/#error_codes_list
+        """
+        kwargs = self.request_params(stream_state=stream_state, stream_slice=stream_slice, next_page_token=next_page_token)
+        return self.api.list(kwargs)
+
+    def read_records(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_slice: Mapping[str, Any] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping[str, Any]]:
```

Comment:
Do we really need this override? At first glance seems that `_send_request` override should be enough

Reply:
Yes, because `ChargebeeStream` class is a child of `Stream` class and in `Stream` class `read_records()` is abstract method.

![image](https://user-images.githubusercontent.com/19872253/127524243-98bdd8d1-4324-4d96-a8ad-0c99f4ce4d96.png)